docs/architecture-decisions/adr005-catalog-core-entities.md

```diff
@@ -0,0 +1,75 @@
+# ADR005: Catalog Core Entities
+
+| Created    | Status |
+| ---------- | ------ |
+| 2020-05-29 | Open   |
+
+## Context
+
+We want to standardize on a few core entities that we are tracking in the Backstage catalog. This allows us to build specific plugins around them.
+
+## Decision
+
+We maintain a catalog of the following core entities:
+
+* **Components** are individual pieces of software
+* **APIs** are the boundaries between different components
+* **Resources** are physical or virtual infrastructure needed to operate a component
+
+![Catalog Core Entities][catalog-core-entities]
+
+### Component
+A component is a piece of software, for example a mobile application feature, web site, backend service or data pipeline (list not exhaustive). A component can be tracked in source control, or use some existing open source or commercial software. It can implement APIs for other components to consume. In turn it might depend on APIs implemented by other components, or resources that are attached to it at runtime.
+
+Component entities are typically defined in YAML descriptor files next to the code of the component, and could look like this (actual schema will evolve):
+```yaml
+apiVersion: backstage.io/v1beta1
+kind: Component
+metadata:
+  name: my-component-name
+spec:
+  type: service
+```
+
+### API
+APIs form an abstraction that allows large software ecosystems to scale. Thus, APIs are a first class citizen in the Backstage model and the primary way to discover existing functionality in the ecosystem.
+
+APIs are implemented by components and make their boundaries explicit. They might be defined using an RPC IDL (eg in Protobuf, GraphQL or similar), a data schema (eg in Avro, TFRecord or similar), or as code interfaces (eg framework APIs in Swift, Kotlin, Java, C++, Typescript etc). In any case, APIs exposed by components need to be in a known machine-readable format so we can build further tooling and analysis on top.
+
+APIs are typically indexed from existing definitions in source control and thus wouldn't need their own descriptor files, but would be stored in the catalog somewhat like this (actual schema will evolve):
+```yaml
+apiVersion: backstage.io/v1beta1
+kind: API
+metadata:
+  name: my-component-api
+spec:
+  type: grpc
+  definition: >
```

Comment:
What do you think about making this a URL that points to the schema instead? I'm a bit worried that this forces you to ingest APIs automatically, you wouldn't type this out by hand and and also store it in a separate .proto file.

Reply:
The proposition mentions that these will typically never be created by hand.

Maintaining a reference to the storage location AS WELL, is probably a good idea either way, in the same way as components that are generated from a file on VCS also should have such a reference.

My current thinking is that there could be an annotation `backstage.io/managed-by: <location-uid>` or similar. If you have a location pointing to a component descriptor, then the ingestor loop would read that file AND all other interesting contextual data, including api declaration files. Either as referenced to by a component annotation, or found nearby in the target structure by convention.

At Spotify, we could find nearby build-info.yaml files and parse them and add contextual information to the component for example.

Reply:
Do you think it's worth adding something here about how we link back from catalog entities to their location in source control? Seems worth documentation, but maybe not super related?

Reply:
We intend to establish docs for a list of well-known pieces of metadata and their semantics - a link like that would be part of that document.

Reply:
Alright, then I assume it's safe to not have it in the ADR.

Reply:
Missed that this was an example of the catalog response not the definition file. I don't really see how we'd be able to generate entities for APIs though? I feel many fields like lifecycle and owner can't really be parsed from the definition files themselves. What would be the logic for determining who owns the API in a `.proto` that appears somewhere in a repo?

Reply:
For example by the component (that is owned) having a ref to where its api definition file is. Or by convention, through the context around where the component file is

Reply:
But doesn't that somewhat defeat the purpose of modelling APIs as separate things to begin with?

Reply:
Well having them indexed and searchable and tied to the producing component and to an owner may be pretty significant in and of itself.

What would you suggest? We could have the option of writing (maybe even scaffolding) descriptor files next to the schema file if one needs more manually curated metadata I guess?

Reply:
I see it as potentially being an end user experience of "oh I just wrote the little yaml file for my service and bam, all of this related info and stuff appeared magically in Backstage, neat" :)

Reply:
I'd actually suggest punting on APIs and Resources for now, just defining the entity kinds and what kinda stuff we want to track, but not try to go further than that yet. I think we've got enough of the problem space covered by just trying to solve modelling of different components, and including more things in the mix won't really expand the picture by much. Then once we have some stuff up for modelling components e2e we'd look at how we want to approach APIs and Resources.

If we did go with APIs now I would just keep it simple and have separate explicit descriptions for APIs. For two main reasons, the first one being that we will need to support it anyway, for when an API needs different metadata than it's implementing component, or it simply isn't implemented by a single component. The other reason being that it gives us a nicer separation of the label and annotation namespacing and docs. Instead of throwing all metadata definitions on components, which will already have more than enough possible labels and annotations, we separate them into more distinct buckets based on what kind of entity they're tied to.

Reply:
Mm we will surely have kinds for APIs and resources, whether they are added by locations or not. We haven't settled on what exact fields each has, and we should indeed be careful not to rush into defining that too soon in too much detail.

But, as our catalog may support anybody ingesting anything by any mechanism they may invent, it becomes not so much a matter of file vs no file, but a matter of timing when we want to flesh out some builtin / blessed ingestion paths right?